src/Transformation/Transformation.php

```diff
@@ -0,0 +1,32 @@
+<?php
+/* Copyright (c) 2017 Richard Klees <richard.klees@concepts-and-training.de> Extended GPL, see docs/LICENSE */
+
+namespace ILIAS\Transformation;
+
+/**
+ * A transformation is a function from one datatype to another.
+ *
+ * It MUST NOT perform any sideeffects, i.e. it must be morally impossible to observe
+ * how often the transformation was actually performed. It MUST NOT touch the providedi
+ * value, i.e. it is allowed to create new values but not to modify existing values.i
+ * This would be an observable sideeffect.  
+ */
+interface Transformation {
+	/**
+	 * Perform the transformation.
+	 *
+	 * @throws \InvalidArgumentException  if the argument could not be transformed
+	 * @param  mixed  $from
+	 * @return mixed
+	 */
+	public function transform($from);
```

Comment:
As far as I know a direct call to the method would be faster then using `__invoke`. It is also more explicit in the code using the transformation. To me, `__invoke` is just a convenience regarding Result::map from #475.